include/xtensor/xmath.hpp

```diff
@@ -1817,36 +1817,42 @@ XTENSOR_INT_SPECIALIZATION_IMPL(FUNC_NAME, RETURN_VAL, unsigned long long);
      * @param axes the axes along which the mean is computed (optional)
      * @return an \ref xexpression
      */
-    template <class E, class X>
+    template <class T = void, class E, class X>
     inline auto mean(E&& e, X&& axes)
     {
+        using value_type = typename std::conditional_t<std::is_same<T, void>::value, double, T>;
```

Comment:
I think using double here is wrong e.g. in the case of complex. Better use big_promote_type

Reply:
Using `big_promote_type` is arguably worse, because you will promote integrals to integrals, even though means are inherently floating point values. Using a `double` here when the user has `complex` is not a default decision (they have to specify it), and I pass in `sum<T>` not `sum<value_type>` which means complex numbers will still be complex.

After some discussion between @JohanMabille and I, we decided that the user should be conscious of the C++ primitive type promotions upon operators like `std::plus<>` and that their results should yield that accordingly, hence why I dropped the use of `big_promote_type` in this case

Reply:
(also, a simple `xt::cast<specified_value_type>(...)` on this functor does not actually yield a `specified_value_type`, which means we might need extra machinery to get the exact return type to work if we were to disregard the primitive type promotion argument)